src/NLog/Targets/DatabaseTarget.cs

```diff
@@ -395,12 +395,13 @@ protected override void Write(LogEventInfo logEvent)
             }
             catch (Exception exception)
             {
+                InternalLogger.Error(exception, "Error when writing to database.");
+
                 if (exception.MustBeRethrown())
                 {
                     throw;
                 }
-
-                InternalLogger.Error("Error when writing to database {0}", exception);
+                
                 this.CloseConnection();
```

Comment:
I think the `if` above must change to `MustBeRethrownImmediately()`, otherwise `CloseConnection()` will not be called in some cases.